pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,139 @@
+	if err != nil {
+		return err
+	}
+	if fullSubnet.SubnetPropertiesFormat == nil {
+		return fmt.Errorf("received nil SubnetProperties when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
+	if fullSubnet.SubnetPropertiesFormat.NetworkSecurityGroup == nil {
+		return fmt.Errorf("received nil NSG when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
+	correctNSGResourceID, err := subnet.NetworkSecurityGroupIDExpanded(architectureVersion, resourcesResourceGroup, subnetName, infraID, isWorkerSubnet)
+	if err != nil {
+		return err
+	}
+	correctNSGResourceID = fmt.Sprintf("/subscriptions/%s/resourceGroups/%s", subscriptionID, correctNSGResourceID)
+
+	if !strings.EqualFold(*fullSubnet.NetworkSecurityGroup.ID, correctNSGResourceID) {
+		r.log.Infof("Fixing NSG from %s to %s", *fullSubnet.NetworkSecurityGroup.ID, correctNSGResourceID)
+		// NSG doesn't match - fixing
+		fullSubnet.NetworkSecurityGroup = &mgmtnetwork.SecurityGroup{ID: &correctNSGResourceID}
+		err = subnetsClient.CreateOrUpdateAndWait(ctx, vnetResourceGroup, vnetName, subnetName, fullSubnet)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) getSubnets(ctx context.Context) (map[subnetDescriptor]bool, string, error) {
+	subnetMap := make(map[subnetDescriptor]bool) // bool is true for master subnets
+	masterResourceGroup := ""
+	// select all workers by the  machine.openshift.io/cluster-api-machine-role: worker label
+	machines, err := r.maocli.MachineV1beta1().Machines(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=worker"})
+	if err != nil {
+		return nil, "", err
+	}
+	for _, machine := range machines.Items {
+		_, subnetDesc, err := r.getDescriptorFromProviderSpec(machine.Spec.ProviderSpec.Value)
+		if err != nil {
+			return nil, "", err
+		}
+		subnetMap[subnetDesc] = false
+	}
+	machines, err = r.maocli.MachineV1beta1().Machines(machineSetsNamespace).List(ctx, metav1.ListOptions{LabelSelector: "machine.openshift.io/cluster-api-machine-role=master"})
+	if err != nil {
+		return nil, "", err
+	}
+	for _, machine := range machines.Items {
+		var subnetDesc subnetDescriptor // declared here due to := rescoping of the masterResourceGroup variable below
+		masterResourceGroup, subnetDesc, err = r.getDescriptorFromProviderSpec(machine.Spec.ProviderSpec.Value)
+		if err != nil {
+			return nil, "", err
+		}
+		subnetMap[subnetDesc] = true
+	}
+	return subnetMap, masterResourceGroup, nil
+}
+
+func (r *AzureNSGReconciler) getDescriptorFromProviderSpec(providerSpec *runtime.RawExtension) (string, subnetDescriptor, error) {
+	var spec azureproviderv1beta1.AzureMachineProviderSpec
+	err := json.Unmarshal(providerSpec.Raw, &spec)
+	return spec.ResourceGroup, subnetDescriptor{
+		resourceGroup: spec.NetworkResourceGroup,
+		vnetName:      spec.Vnet,
+		subnetName:    spec.Subnet,
+	}, err
```

Comment:
If unMarshall fails better not to return. These might be not pointers but easier to read if you add return and error handling separately. (I know I asked to do this above (one liner)) But this compromises readability, would you not agree?

Reply:
Refactored to a separate if, returning pointers instead of objects, and returning nil if there's an error instead of actual values.